dhcp-server: save bound leases to persistent storage, and load it on start

yuwata opened the pull request:
Closes #29991.
Fixes #31708.

yuwata commented:
Finally, it is ready for review. PTAL!

poettering commented on src/network/networkd-dhcp-server.c:
```diff
@@ -522,6 +522,14 @@ static int dhcp4_server_configure(Link *link) {
         }
 
+        _cleanup_free_ char *lease_file = NULL;
+        if (asprintf(&lease_file, "/var/lib/systemd/network/dhcp-server-lease/%s", link->ifname) < 0)
```
path_join()?

poettering commented on src/network/networkd.c:
```diff
@@ -69,17 +70,15 @@ static int run(int argc, char *argv[]) {
+                       "/run/systemd/netif/links/",
+                       "/run/systemd/netif/leases/",
+                       "/run/systemd/netif/lldp/",
+                       "/var/lib/systemd/network/dhcp-server-lease/") {
```
hmm, so networkd is an early boot service, i.e. might be running super early, when /var/ is not mounted yet.

so i think we probably shouldn't bother with creating the dir here. I mean, we only create this because of inotify, right, but people should probably continue to watch /run/systemd/netif/leases/ for that, no? after all the data in that dir and in "/var/lib/systemd/network/dhcp-server-lease/" is pretty much the same, except the latter is in json, right?

poettering replied:
oh, hmm, so othe other dir is for the client side, right, and this is for the server side?

so they might actually watch this... hmm, dunno then.

not sure what to do about this...

poettering commented:
so looks good in general, i just would like some care taken that we work sensibly in an environment with /var/ being still absent.

poettering commented:
i do wonder how useful this is going to be IRL, i.e. given that we are an early boot service, possibly already running in the initrd, where we can't really load the lease, even if we wanted... I am not against adding this though, since it sounds useful for interfaces triggered later though.

poettering commented:
so, to be fully correct we would have to tell networkd when /var/ is around, similar to how we do it for journald. And then later, tell it when /var/ is gone again.

And then, when networkd is told about /var/ being ready it would have to reread all leases...

I have the suspicion that it would be nicest of networkd would keep a primary set of leases under /run/, and only treat the stuff in /var/ as a secondary copy of it. So that people would inotify watch the version in /run, not the one in /var.

keszybz commented on src/libsystemd-network/sd-dhcp-server.c:
```diff
@@ -1565,3 +1581,26 @@ int sd_dhcp_server_set_relay_agent_information(
+        assert_return(server, -EINVAL);
+        assert_return(path, -EINVAL);
+
+        if (isempty(path)) {
```
It's strange to accept `""` here. I'd just say `if (!path)`.

keszybz commented on src/libsystemd-network/sd-dhcp-server.c:
```diff
@@ -1565,3 +1581,26 @@ int sd_dhcp_server_set_relay_agent_information(
+                return -EINVAL;
+
+        r = free_and_strdup(&server->lease_file, path);
+        if (r < 0)
+                return r;
+        if (r == 0)
```
`if r <= 0` ?

keszybz reviewed (changes requested):
Looks good in general. I think we can merge this as is, since it's already useful. As Lennart said, it doesn't cover the case where networkd is started very early, but for many systems this is not an issue. We can figure out how to deal with those cases later (if at all).

keszybz commented:
> And then, when networkd is told about /var/ being ready it would have to reread all leases...

I don't think that'd be useful. If networkd is told to configure an interface, by the time we have mounted `/var`, it'll most likely have already reacquired the lease. I see two possibilities: for some interfaces, we mark the interface as "requiring /var" and let networkd delay the configuration of that interface. Yu rewrote the networkd internals to allow asynchronous operation, so this actually might be feasible to implement cleanly. A different approach would be to push the storage for certain leases to e.g. an EFI variable, so that it becomes available very early in boot. But I'm not sure if that's worth the trouble.

yuwata commented:
@poettering @keszybz Thank you for the review.
In the revised version, a minimal subset that @poettering requested is introduced. Please take a look!

yuwata commented:
@mrc0mmand systemd-networkd-dhcp-server.service will not be installed on CentOS CI (Arch Linux + sanitizers) ??

(TODO; also update test-networkd.py)

mrc0mmand commented:
> @mrc0mmand systemd-networkd-dhcp-server.service will not be installed on CentOS CI (Arch Linux + sanitizers) ??

Fixed (and restarted the affected job). I think it would make sense for systemd-networkd-tests.py to copy over necessary units files from the build dir to `/run` if `--build-dir=` is specified (instead of just overriding paths in `ExecStart=`), so such changes get picked up automagically. I'll add that to my TODO list.

poettering commented on src/libsystemd-network/sd-dhcp-server-lease.c:
```diff
@@ -204,3 +204,69 @@ int sd_dhcp_server_set_static_lease(
+        return json_variant_append_arrayb(
+                        array,
+                        JSON_BUILD_OBJECT(
+                                        JSON_BUILD_PAIR_BYTE_ARRAY(
```
btw, why not use JSON_BUILD_BASE64() for this? I think this is actually more in line how most web people would encode binary blobs.

poettering replied:
(i.e. for new formats i'd always prefer base64 strings in stead of byte arrays to encode binary blobs on json)

poettering replied:
ahh, this is not a new format i guess... then ignore my comment

poettering commented on src/network/networkctl.c:
```diff
@@ -119,7 +119,7 @@ static int check_netns_match(void) {
         if (id == 0) {
                 log_debug("systemd-networkd.service not running in a network namespace (?), skipping netns check.");
-                return 0;
+                goto pass;
```
why use goto? insead you could have a else branch, below.

i mean, using goto is ok to use to jump downwards, but also: if we can avoid it easily, i'd still avoid it.

poettering commented:
hmm, i am a bit concerned about making this a per-interface thing and a per-dhcp thing. I don't see the point of making it so fine grained. After all we really just care about two global states: before /var/ is mounted, and after /var/ is mounted. starting individual dhcp servers on individual interfaces sounds needlessly complex to me.

the way i though this would work is if there'd be a new boolean field in struct `Manager` called `persistent_storage_ready` or so. And the service would just toggle that bool via varlink, once during boot and once during shutdown. And all dhcp servers would react to this simultaneously the same way.

and the nice thing would be that this is generic, so if we one day we have other resources we need to persist (for example, it might make sense to also save acquired leases to disk in some scenarios, in the dhcp client, or any other info in fact that takes time to acquire and that would make sense to keep around for quick turnarounds, outside of the realm of dhcp) we have a generic knob already for that.

poettering commented on src/libsystemd-network/sd-dhcp-server.c:
```diff
@@ -1341,6 +1355,10 @@ int sd_dhcp_server_start(sd_dhcp_server *server) {
 
+        r = dhcp_server_load_leases(server);
+        if (r < 0)
+                log_dhcp_server(server, "Failed to load lease file %s, ignoring: %m", strna(server->lease_file));
```
btw, if we are offline for a long time during reboot, we might need to immediately terminate these leases? is there code in place for that already?

yuwata commented:
@mrc0mmand 
> I think it would make sense for systemd-networkd-tests.py to copy over necessary units files from the build dir to `/run` if `--build-dir=` is specified (instead of just overriding paths in `ExecStart=`),

Thank you for the suggestion. Done.

@poettering Addressed all comments. PTAL!

poettering commented on src/libsystemd-network/dhcp-server-lease-internal.h:
```diff
@@ -20,15 +20,15 @@ typedef struct sd_dhcp_server_lease {
         be32_t gateway;
         uint8_t chaddr[16];
-        usec_t expiration;
+        triple_timestamp expiration;
```
hmm, i think we should only keep the timestamp in CLOCK_BOOTTIME in memory, but when writing it out convert it to CLOCK_REALTIME and write it out too in that format then.

Why? simply because dhcp expiration timeouts can be super long, and systems sometimes come up with bad clocks. hence we should only convert CLOCK_BOOTTIME to CLOCK_REALTIME at the moment we really have to, i.e. when writing, but never before. If that makes any sense?

poettering commented on src/libsystemd-network/sd-dhcp-server-lease.c:
```diff
@@ -215,7 +215,8 @@ static int dhcp_server_lease_append_json(sd_dhcp_server_lease *lease, JsonVarian
                                  JSON_BUILD_PAIR_IN4_ADDR_NON_NULL("Address", &(struct in_addr) { .s_addr = lease->address }),
                                  JSON_BUILD_PAIR_STRING_NON_EMPTY("Hostname", lease->hostname),
-                                 JSON_BUILD_PAIR_FINITE_USEC("ExpirationUSec", lease->expiration)));
+                                 JSON_BUILD_PAIR_FINITE_USEC("ExpirationUSec", lease->expiration.boottime),
```
ithink it would make sense to store the boot id here too, i.e. sd_id128_get_boot(), after all without it it's very hard to make sense of the boottime clock. but if you store the boot id as well, then it's very easy for people to figure out if they can use the boot time, or have to use the wallclock timestamp

poettering commented on src/libsystemd-network/sd-dhcp-server-lease.c:
```diff
@@ -215,7 +215,8 @@ static int dhcp_server_lease_append_json(sd_dhcp_server_lease *lease, JsonVarian
                                  JSON_BUILD_PAIR_STRING_NON_EMPTY("Hostname", lease->hostname),
-                                 JSON_BUILD_PAIR_FINITE_USEC("ExpirationUSec", lease->expiration)));
+                                 JSON_BUILD_PAIR_FINITE_USEC("ExpirationUSec", lease->expiration.boottime),
+                                 JSON_BUILD_PAIR_FINITE_USEC("ExpirationTimestamp", lease->expiration.realtime)));
```
I#d call this ExpirationRealtimeUSec, i.e. emphasize what this is, and linux/posix calls this CLOCK_REALTIME after all.

poettering commented on src/network/networkd-manager-varlink.c:
```diff
@@ -164,6 +166,37 @@ static int vl_method_get_lldp_neighbors(Varlink *vlink, JsonVariant *parameters,
+        r = varlink_verify_polkit_async(
+                                vlink,
+                                manager->bus,
+                                "io.systemd.Network.SetPersistentStorage",
```
hmm, we so far always spelled the policykit actions we define in lowercase with dashes?

poettering commented on units/systemd-networkd-persistent-storage.service:
```diff
@@ -0,0 +1,24 @@
+After=systemd-networkd.service
+BindsTo=systemd-networkd.service
+ConditionPathExists=!/etc/initrd-release
+
```
hmm, i think this should run during early boot already (i.e. have DefaultDependencies=no), but order itself after systemd-remount-fs.service and RequiresMountsFor=/var/lib/systemd/network, so that it is not called before that dir is there and writable.

YHNdnzj commented on man/networkctl.xml:
```diff
@@ -495,6 +495,21 @@ s - Service VLAN, m - Two-port MAC Relay (TPMR)
+          <para>Notify <filename>systemd-networkd.service</filename> if the persistent storage for the
+          service is ready. This is called by
+          <filename>systemd-networkd-persistent-storage.service</filename>, and is not necessary to be called
+          by users or administrators.</para>
```
Hmm, if this is the case, maybe we should just not document it? This could still give wrong impressions to users if they only check --help, and as said it's generally not expected that it shall be called outside of the service.

YHNdnzj replied:
Or even make this an `invoked_as` thing?

YHNdnzj replied:
`switch-root` is pretty much generic. It takes a path although we assume `/sysroot/` across our code. People who want to use a different set of initrd units could and do make use of it (I once saw people using it for achieving something similar to `soft-reboot`. There's a reason why we restricted switch-root to be used in initrd only _after_ `soft-reboot` was introduced)

This, however, only controls an internal state of networkd. And I really cannot think of a case where people would call it behind our back.

> Otherwise, people needs to read code to understand what the command does.

Hmm, we have tools like `systemd-creds`, that when invoked through socket activation, automatically sets up a varlink server. This is nowhere similar to their CLI usage.

YHNdnzj replied:
That said, that's only an opinion. I'm not strong against this if you want to do it anyway :)

yuwata replied:
If not only you suggest that we should hide the command from the document, then I think we should use varlinkctl in the .service unit in this case, instead of introducing such tiny wrapper of the varlink command.

yuwata commented:
@poettering I hope all your comments are addressed. PTAL!

evverx commented on src/network/networkd-json.c:
```diff
@@ -968,11 +907,11 @@ static int dhcp_server_append_json(Link *link, JsonVariant **v) {
                 return r;
 
-        r = dhcp_server_offered_leases_append_json(link, &w);
+        r = dhcp_server_bound_leases_append_json(link->dhcp_server, &w);
```
With a bunch of (nonsensical) leases I got a weird error message:
```sh
networkctl status 15 --json pretty 
Failed to get description: Structure needs cleaning
```
```sh
Got message type=method_call sender=:1.14972 destination=org.freedesktop.network1 path=/org/freedesktop/network1 interface=org.freedesktop.network1.Manager member=Describe  cookie=2 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Failed to build JSON data: Structure needs cleaning
Sent message type=error sender=n/a destination=:1.14972 path=n/a interface=n/a member=n/a cookie=56 reply_cookie=2 signature=s error-name=System.Error.EUCLEAN error-message=Structure needs cleaning
Failed to process message type=method_call sender=:1.14972 destination=org.freedesktop.network1 path=/org/freedesktop/network1 interface=org.freedesktop.network1.Manager member=Describe cookie=2 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a: Structure needs cleaning
```
`--json=off` seems to be fine:
```sh
networkctl status 15 --json off
...
         Offered DHCP leases: 192.168.18.156 (to ETHER)
                              192.168.18.152 (to ETHER)
                              192.168.18.158 (to ETHER)
                              192.168.18.150 (to ARP/LL)
                              192.168.18.148 (to ARP/LL)
                              192.168.18.151 (to ETHER)
                              192.168.18.153 (to ETHER)
                              192.168.18.146 (to ETHER)
                              192.168.18.157 (to ETHER)
                              192.168.18.154 (to ETHER)
                              192.168.18.149 (to ARP/LL)
                              192.168.18.155 (to ETHER)
                              192.168.102.171 (to IAID:0x5888c0c1/DUID)
                              192.168.18.147 (to ARP/LL)
```
When `networkd` was restarted 4 leases were restored:
```
         Offered DHCP leases: 192.168.18.149 (to ARP/LL)
                              192.168.18.147 (to ARP/LL)
                              192.168.18.155 (to ETHER)
                              192.168.102.171 (to IAID:0x5888c0c1/DUID)
```
and the `json` part worked too
```
networkctl status 15 --json pretty | jq .DHCPServer.Leases[].Address
[
  192,
  168,
  18,
  149
]
[
  192,
  168,
  18,
  147
]
[
  192,
  168,
  18,
  155
]
[
  192,
  168,
  102,
  171
]
```

evverx replied:
Looks like it's caused by non UTF-8 hostnames:
```python
>>> rdpcap('DHCP-non-utf-8-hostname')[0][DHCP].options
[('message-type', 1),
 ('client_id', b'\xffX\x88\xc0\xc1\x00\x02\x00\x00\xab\x11G\xa3]\xec\x905E9'),
 ('param_req_list', [1, 3, 6, 12, 15, 33, 42, 101, 114, 120, 121]),
 ('max_dhcp_size', 1472),
 ('rapid_commit', b''),
 ('hostname', b'\xff\xfe'),
 'end']
```
```
DHCPv4 server: DISCOVER (0x3c7950a8)
DHCPv4 server: ACK (0x3c7950a8)
DHCPv4 server: Failed to save leases, ignoring: Structure needs cleaning
```

yuwata replied:
Thanks. Should be fixed now. Please test again.

evverx replied:
I'll take a look a bit later.

Having experimented with this PR I think it would be great to make it easier to clean up leases. Previously to get rid of nonsensical leases it was possible to just restart networkd. With this PR applied networkd has to be stopped and the`/var/lib/systemd/network/dhcp-server-lease/*` files have to be removed manually as far as I can tell.

Also based on my experiments networkd doesn't react to DHCP packets at all until the storage is ready. I'm not sure it's necessary when the DHCP server just relays packets. I'm not sure it should stop handing out leases with no storage either. My understanding is that the idea was to hand out leases (regardless of whether the storage is ready or not) and then save it once the "persistent storage" is ready. Dunno. I haven't fully read the thread yet.

yuwata replied:
>  I'm not sure it's necessary when the DHCP server just relays packets.

Good point! Yeah, it is not necessary to wait for the persistent storage when the server is running in relay mode.

yuwata replied:
> I'm not sure it should stop handing out leases with no storage either. My understanding is that the idea was to hand out leases (regardless of whether the storage is ready or not) and then save it once the "persistent storage" is ready. Dunno. I haven't fully read the thread yet.

If /var/lib/systemd/network is on a non-root partition, then we need to wait the partition is mounted. Otherwise, on reboot, the DHCP server may start before the partition is mounted, hence we cannot load previously bound leases, and an address already assigned to a host may be bound to another host.
So, we need to read the lease file before starting the DHCP server, and to support that we need to wait for the persistent storage being ready.

yuwata replied:
> Having experimented with this PR I think it would be great to make it easier to clean up leases. Previously to get rid of nonsensical leases it was possible to just restart networkd. With this PR applied networkd has to be stopped and the`/var/lib/systemd/network/dhcp-server-lease/*` files have to be removed manually as far as I can tell.

Right. So, I also updated the our test script, to remove the persistent storage before starting networkd cleanly.

evverx replied:
Nonsensical leases aside I don't think it should be necessary to manually remove leases when containers are restarted for example. With (default) nspawn containers getting their IPs from networkd on the host `networkctl` looks really weird after those containers get rebooted. Below `192.168.46.129` gets mixed up with `192.168.249.177`.

The first run looks fine:
```
                     Address: 192.168.249.177
                              169.254.20.255
                              fe80::741a:afff:fe88:1910
           Activation Policy: up
         Required For Online: yes
                Connected To: C on port host0
         Offered DHCP leases: 192.168.249.182 (to ARP/LL)
                              192.168.249.183 (to ETHER)
                              192.168.249.180 (to ARP/LL)
                              192.168.249.187 (to ETHER)
                              192.168.249.179 (to ARP/LL)
                              192.168.249.185 (to ETHER)
                              192.168.249.178 (to ETHER)
                              192.168.249.189 (to ETHER)
                              192.168.249.184 (to IAID:0x5888c0c1/DUID)
                              192.168.249.188 (to ETHER)
                              192.168.249.186 (to ETHER)
                              192.168.249.181 (to ARP/LL)
                              192.168.249.190 (to ETHER)
```
The second run after the container was rebooted looks weird:
```
                     Address: 192.168.46.129
                              fe80::741a:afff:fe88:1910
           Activation Policy: up
         Required For Online: yes
                Connected To: C on port host0
         Offered DHCP leases: 192.168.249.178 (to ETHER)
                              192.168.249.181 (to ARP/LL)
                              192.168.249.189 (to ETHER)
                              192.168.249.187 (to ETHER)
                              192.168.249.180 (to ARP/LL)
                              192.168.249.182 (to ARP/LL)
                              192.168.249.188 (to ETHER)
                              192.168.249.186 (to ETHER)
                              192.168.249.185 (to ETHER)
                              192.168.249.183 (to ETHER)
                              192.168.249.179 (to ARP/LL)
                              192.168.46.136 (to IAID:0x5888c0c1/DUID)
                              192.168.249.190 (to ETHER)
```

evverx replied:
The "Structure needs cleaning" issue is gone as far as I can see.

yuwata replied:
Interesting. But that's intended. Any leases should be dropped only when it is expired.
BTW, when picking up an IPv4 address requested with a null address, e.g. 0.0.0.0/24, I think we should take into account the lease file.

keszybz commented on test/test-network/systemd-networkd-tests.py:
```diff
@@ -7482,45 +7471,22 @@ def test_mtu_link_ipv6_mtu(self):
-        raise RuntimeError(f"{systemd_source_dir} doesn't appear to be a systemd source tree")
+        source_dir = os.path.normpath(os.path.join(os.path.dirname(os.path.abspath(__file__)), "../../"))
+    if not os.path.exists(os.path.join(source_dir, "meson_options.txt")):
+        raise RuntimeError(f"{source_dir} doesn't appear to be a systemd source tree")
```
Preexisting condition, but `RuntimeError` is a javaism that is not useful in Python code…

yuwata replied:
Then what should we use?

keszybz replied:
A normal exception, ValueError or some other standard exception or maybe even AssertionError, i.e. replace the whole conditional with `assert`.

keszybz commented on src/network/org.freedesktop.network1.policy:
```diff
@@ -183,4 +183,15 @@
         </action>
 
+        <action id="org.freedesktop.network1.set-persistent-storage">
+                <description gettext-domain="systemd">Set if persistent storage for systemd-networkd is ready</description>
```
"Set if" is not a valid grammatical construct.
Maybe "Specify whether persistent storage for systemd-networkd is available."

keszybz commented on src/network/org.freedesktop.network1.policy:
```diff
@@ -183,4 +183,15 @@
 
+        <action id="org.freedesktop.network1.set-persistent-storage">
+                <description gettext-domain="systemd">Set if persistent storage for systemd-networkd is ready</description>
+                <message gettext-domain="systemd">Authentication is required to set if persistent storage for systemd-networkd is ready.</message>
```
Also here.

keszybz commented on man/networkctl.xml:
```diff
@@ -495,6 +495,21 @@ s - Service VLAN, m - Two-port MAC Relay (TPMR)
+          <replaceable>BOOL</replaceable>
+        </term>
+        <listitem>
+          <para>Notify <filename>systemd-networkd.service</filename> if the persistent storage for the
```
"if" → "that"

keszybz commented on man/networkctl.xml:
```diff
@@ -495,6 +495,21 @@ s - Service VLAN, m - Two-port MAC Relay (TPMR)
+        <listitem>
+          <para>Notify <filename>systemd-networkd.service</filename> if the persistent storage for the
+          service is ready. This is called by
+          <filename>systemd-networkd-persistent-storage.service</filename>, and is not necessary to be called
```
"and is not necessary to be called by …" is very awkward.

keszybz commented on units/systemd-networkd-persistent-storage.service:
```diff
@@ -0,0 +1,27 @@
+#  (at your option) any later version.
+
+[Unit]
+Description=Notify if the persistent storage for systemd-networkd.service is ready
```
Descriptions must be "Title Case" and work well with "Starting …" and "Started …".
"Notify systemd-networkd.service that Persistent Storage is Available" ?
Or maybe even "Enable Persistent Storage in systemd-networkd"?

keszybz commented on src/basic/time-util.h:
```diff
@@ -77,6 +77,7 @@ typedef enum TimestampStyle {
 usec_t now(clockid_t clock);
 nsec_t now_nsec(clockid_t clock);
 
+usec_t map_clock_usec_internal(usec_t from, usec_t from_base, usec_t to_base);
 usec_t map_clock_usec(usec_t from, clockid_t from_clock, clockid_t to_clock);
```
It's strange to make function called "…_internal" public. It's also not clear how the two functions differ.
I think ` map_clock_usec_internal` should be renamed to `map_clock_usec_raw` or something like that.

keszybz commented on src/libsystemd-network/sd-dhcp-server-lease.c:
```diff
@@ -219,21 +219,33 @@ static int dhcp_server_lease_append_json(sd_dhcp_server_lease *lease, JsonVarian
                 if (r < 0)
                         return r;
 
+                usec_t exp_r = map_clock_usec_internal(lease->expiration, now_b, now_r);
```
Hmm, why is `map_clock_usec` not appropriate? Please add a comment.

keszybz commented on src/libsystemd-network/sd-dhcp-server.c:
```diff
@@ -1565,3 +1583,18 @@ int sd_dhcp_server_set_relay_agent_information(
+        assert_return(server, -EINVAL);
+        assert_return(!sd_dhcp_server_is_running(server), -EBUSY);
+
+        if (!path) {
+                server->lease_file = mfree(server->lease_file);
+                return 0;
+        }
+
+        if (!path_is_safe(path))
+                return -EINVAL;
```
Please change this to `if (path && !path_is_safe(path))` and drop the separate handling of `if (!path)`. This way, we get the expected return value 1 if the value changes, and 0 otherwise. (And the code is simpler.)

keszybz commented on src/network/networkd-dhcp-server.c:
```diff
@@ -143,6 +144,55 @@ int network_adjust_dhcp_server(Network *network, Set **addresses) {
+        if (!link_has_carrier(link))
+                return 0;
+
+        /* TODO: Maybe, also check the system time is synced. If the system does not have RTC battery, then
+         * the realtime clock in not usable in the early boot stage, and all saved leases may be wrongly
+         * handled as expired and dropped. */
```
I guess we'll want to order `systemd-networkd-persistent-storage.service` after `time-sync.target`? But let's not do this now, this PR is complicated enough already.

poettering replied:
yeah, sounds like a job for time-sync.target

evverx commented on src/libsystemd-network/fuzz-dhcp-server.c:
```diff
@@ -62,9 +65,11 @@ static int add_static_lease(sd_dhcp_server *server, uint8_t i) {
 }
 
 int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+        _cleanup_(rmdir_and_freep) char *tmpdir = NULL;
```
I think it should be `rm_rf_physical_and_freep` instead. With `rmdir_and_freep` the fuzz target doesn't remove temporary lease files and fills up `/tmp`.

evverx commented on src/libsystemd-network/sd-dhcp-server-lease.c:
```diff
@@ -93,7 +98,10 @@ int dhcp_server_set_lease(sd_dhcp_server *server, be32_t address, DHCPRequest *r
+        /* Currently, the hostname field in the request is only used when dumping leases in JSON format, and
+         * the field is optional. So, let's silently ignore non UTF8 hostname. This needs to be updated when
+         * we really use the hostname field. */
+        if (req->hostname && utf8_is_valid(req->hostname)) {
```
Interestingly the "Structure needs cleaning" thing was apparently introduced in https://github.com/systemd/systemd/pull/27817 and is reproducible with the main branch. I'm not sure why UTF-8 should be expected here. I wonder if it would make sense to dump bytes or something like that instead of using `JSON_BUILD_PAIR_STRING_NON_EMPTY` (where UTF-8 is enforced)

(It can probably be done later though)

evverx replied:
I opened https://github.com/systemd/systemd/issues/31708 so as not to forget about this.

yuwata replied:
Thanks. Maybe, we should also add `HostnameRaw` field.

evverx replied:
The part loading the DHCP client leases apparently enforces UTF-8 too when it loads its leases (but it doesn't do it when it saves them):
```sh
# networkctl status host0 --no-pager                                                                                                                 
/run/systemd/netif/leases/2:13: invalid UTF-8 value for key ROOT_PATH: '��', ignoring.
...
```
I'm not exactly sure what to do about it but I can see that `fuzz-dhcp-server` now tries to save and then load its leases so the server part should be tested with all sorts of weird leases at least.

I'm not sure why CIFuzz is green but `fuzz-dhcp-server` fails to load the following lease (I'm guessing the hostname is the culprit there too):
```sh
{
  "BootID": "8fb3b8fae02a47adbe4052689d36652b",
  "Leases": [
    {
      "ClientId": [
        1,
        0,
        0,
        127,
        93,
        0,
        0
      ],
      "Address": [
        10,
        0,
        0,
        58
      ],
      "Hostname": "g\u0002=\u0002\u007f\u0002\u0002f\u0002\u0002e",
      "ExpirationUSec": 278876468747,
      "ExpirationRealtimeUSec": 1710103386715862
    },
    {
      "ClientId": [
        2,
        3
      ],
      "Address": [
        10,
        0,
        0,
        3
      ],
      "ExpirationUSec": 361676468686,
      "ExpirationRealtimeUSec": 1710186186715801
    },
    {
      "ClientId": [
        2,
        2
      ],
      "Address": [
        10,
        0,
        0,
        2
      ],
      "ExpirationUSec": 361676468605,
      "ExpirationRealtimeUSec": 1710186186715720
    }
  ]
}
```

yuwata replied:
@evverx I have no idea about that. Maybe, there were garbage? Do you reproduce the issue reliably?

evverx replied:
The fuzz target crashes reliably. I sent a DHCPDISCOVER message with that hostname
```sh
    Option: (57) Maximum DHCP Message Size
        Length: 2
        Maximum DHCP Message Size: 1472
    Option: (80) Rapid commit
        Length: 0
        Value: <MISSING>
    Option: (12) Host Name
        Length: 11
        Host Name: g\002=\002\177\002\002f\002\002e
    Option: (255) End
        Option End: 255
```
and it broke `networkctl` at least:
```sh
networkctl status 30 --json=pretty
Failed to parse JSON: Invalid argument

networkctl status 30 |& grep -i lease
         Offered DHCP leases: 192.168.15.82 (to IAID:0x5888c0c1/DUID)
```
When networkd was restarted the lease got lost:
```
networkctl status 30 |& grep -i lease
         Offered DHCP leases: none
```
```
DHCPv4 server: Failed to load lease file /var/lib/systemd/network/dhcp-server-lease/ve-TEST-01, ignoring: Success
```

evverx replied:
For the record it's valid UTF-8 so it's different from https://github.com/systemd/systemd/issues/31708

yuwata replied:
Found the root cause. Now the issue should be fixed.

evverx replied:
Looks like it's gone.

(Apparently it's reproducible with the main branch too so it wasn't introduced in this PR either).

keszybz commented on src/libsystemd-network/sd-dhcp-server.c:
```diff
@@ -1565,3 +1583,13 @@ int sd_dhcp_server_set_relay_agent_information(
+        assert_return(server, -EINVAL);
+        assert_return(!sd_dhcp_server_is_running(server), -EBUSY);
+
+        if (!path || !path_is_safe(path))
```
This condition is wrong: it should be `if (path && !path_is_safe(path))`, because we want to allow NULL to unset the lease file.

evverx commented:
I'm not sure how the list of leases should be kept up to date. For example I'm not sure why I have 44 leases when the default nspawn container should have 13 IP addresses tops:
```sh
networkctl status 35 --json short | jq '.DHCPServer.Leases | length'
44
```
and what should happen when I start renewing all those leases. I guess I can remove the lease file manually but it looks a bit kludgy to me. Then again that use case isn't normal. Dunno.

yuwata commented:
> I'm not sure how the list of leases should be kept up to date. For example I'm not sure why I have 44 leases when the default nspawn container should have 13 IP addresses tops:
> 
> ```shell
> networkctl status 35 --json short | jq '.DHCPServer.Leases | length'
> 44
> ```
> 
> and what should happen when I start renewing all those leases. I guess I can remove the lease file manually but it looks a bit kludgy to me. Then again that use case isn't normal. Dunno.

That should be happens when the server change to use the address range, e.g. previously 192.168.0.0/24 but now 192.168.1.0/24. Even in such case, as already said, I intentionally keep the previous leases in 192.168.0.0/24, as the address range change may be misconfiguration, and the server may soon restarted again with the previous range.

I'd like to merge the PR as is, and will submit follow-up commit to improve the situation of running DHCP server with Address=0.0.0.0/24 in container.

yuwata commented:
CI failures in Ubuntu noble are in homed, unrelated. Merging.

yuwata commented:
Please backport fa3357b9e8d9d7a486902d0b6d4b4015fc10aac0 to stable.

poettering commented on src/shared/varlink-io.systemd.Network.c:
```diff
@@ -40,11 +40,16 @@ static VARLINK_DEFINE_METHOD(
+static VARLINK_DEFINE_METHOD(
+                SetPersistentStorage,
+                VARLINK_DEFINE_INPUT(Ready, VARLINK_BOOL, 0));
+
```
might be nice to do:

```c
static VARLINK_DEFINE_ERROR(StorageIsReadOnly);
```

here, and then list in in the interface, and then fire that in the EROFS case above

poettering commented on src/network/networkd-manager-varlink.c:
```diff
@@ -164,6 +167,53 @@ static int vl_method_get_lldp_neighbors(Varlink *vlink, JsonVariant *parameters,
+                if (r < 0)
+                        return log_warning_errno(r, "Failed to check if /var/lib/systemd/network/ is writable: %m");
+                if (r > 0)
+                        return log_warning_errno(SYNTHETIC_ERRNO(EROFS), "The directory /var/lib/systemd/network/ is read-only.");
```
maybe fire a proper varlink error here, i.e. `return varlink_error(vlink, "io.systemd.Network.StorageReadOnly", NULL);`

you get the idea

poettering reviewed (commented):
looks great, some minor post-comit suggestions

mrc0mmand commented:
Hmm, looks like that with this PR the boot on C9S gets stuck sometimes, i.e.:
 - https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-centos9s/1530/consoleFull
 - https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-centos9s/1516//console
 - https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-centos9s/1519/console
 - ...

I'm trying to reproduce this locally to see what's going on.

mrc0mmand commented:
> Hmm, looks like that with this PR the boot on C9S gets stuck sometimes, i.e.:
> 
>     * https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-centos9s/1530/consoleFull
> 
>     * https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-centos9s/1516//console
> 
>     * https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-centos9s/1519/console
> 
>     * ...
> 
> 
> I'm trying to reproduce this locally to see what's going on.

Opened https://github.com/systemd/systemd/issues/31742 with a journal from one such machine.
